vpr/src/device/rr_graph/rr_graph_obj.cpp

```diff
@@ -0,0 +1,1753 @@
+
+    if (DEC_DIRECTION == node_direction(node)) {
+        end_coordinator.set(node_xlow(node), node_ylow(node));
+    }
+
+    return end_coordinator;
+}
+
+short RRGraph::node_fan_in(RRNodeId node) const {
+    return node_in_edges(node).size();
+}
+
+short RRGraph::node_fan_out(RRNodeId node) const {
+    return node_out_edges(node).size();
+}
+
+short RRGraph::node_capacity(RRNodeId node) const {
+    VTR_ASSERT_SAFE(valid_node_id(node));
+    return node_capacities_[node];
+}
+
+short RRGraph::node_ptc_num(RRNodeId node) const {
+    VTR_ASSERT_SAFE(valid_node_id(node));
+    return node_ptc_nums_[node];
+}
+
+short RRGraph::node_pin_num(RRNodeId node) const {
+    VTR_ASSERT_MSG(node_type(node) == IPIN || node_type(node) == OPIN,
+                   "Pin number valid only for IPIN/OPIN RR nodes");
+    return node_ptc_num(node);
+}
+
+short RRGraph::node_track_num(RRNodeId node) const {
+    VTR_ASSERT_MSG(node_type(node) == CHANX || node_type(node) == CHANY,
+                   "Track number valid only for CHANX/CHANY RR nodes");
+    return node_ptc_num(node);
+}
+
+short RRGraph::node_class_num(RRNodeId node) const {
+    VTR_ASSERT_MSG(node_type(node) == SOURCE || node_type(node) == SINK, "Class number valid only for SOURCE/SINK RR nodes");
+    return node_ptc_num(node);
+}
+
+short RRGraph::node_cost_index(RRNodeId node) const {
+    VTR_ASSERT_SAFE(valid_node_id(node));
+    return node_cost_indices_[node];
+}
+
+e_direction RRGraph::node_direction(RRNodeId node) const {
+    VTR_ASSERT_SAFE(valid_node_id(node));
+    VTR_ASSERT_MSG(node_type(node) == CHANX || node_type(node) == CHANY, "Direction valid only for CHANX/CHANY RR nodes");
+    return node_directions_[node];
+}
+
+const char* RRGraph::node_direction_string(RRNodeId node) const {
+    VTR_ASSERT_SAFE(valid_node_id(node));
+    VTR_ASSERT_MSG(node_type(node) == CHANX || node_type(node) == CHANY, "Direction valid only for CHANX/CHANY RR nodes");
+
+    if (INC_DIRECTION == node_direction(node)) {
+        return "INC_DIR";
+    } else if (DEC_DIRECTION == node_direction(node)) {
+        return "DEC_DIR";
+    } else if (BI_DIRECTION == node_direction(node)) {
+        return "BI_DIR";
+    }
+
+    VTR_ASSERT(NO_DIRECTION == node_direction(node));
+    return "NO_DIR";
+}
+
+e_side RRGraph::node_side(RRNodeId node) const {
+    VTR_ASSERT_SAFE(valid_node_id(node));
+    VTR_ASSERT_MSG(node_type(node) == IPIN || node_type(node) == OPIN, "Side valid only for IPIN/OPIN RR nodes");
+    return node_sides_[node];
+}
+
+const char* RRGraph::node_side_string(RRNodeId node) const {
+    VTR_ASSERT_SAFE(valid_node_id(node));
+    VTR_ASSERT_MSG(node_type(node) == IPIN || node_type(node) == OPIN, "Side valid only for IPIN/OPIN RR nodes");
+    return SIDE_STRING[node_side(node)];
+}
+
+/* Get the resistance of a node */
+float RRGraph::node_R(RRNodeId node) const {
+    VTR_ASSERT_SAFE(valid_node_id(node));
+    return node_Rs_[node];
```

Comment:
FYI, the current RR graph implementation fly-weights the R/C values. In the future we could do the same here, but for now its fine to leave this as is.

Reply:
Got it. I may remove it when start constructing the rr_graph builders